subsys/bluetooth/host/audio/bap.c

```diff
@@ -20,84 +20,1323 @@
+
+	/* TODO: Verify state changes with keeping in mind that broadcast
+	 * endpoints have different state changes than unicast
+	 */
+
+	old_state = ep->status.state;
+	ep->status.state = state;
+
+	if (!ep->stream || old_state == state) {
+		return;
+	}
+
+	if (state == BT_AUDIO_EP_STATE_IDLE) {
+		bt_audio_stream_detach(ep->stream);
+	}
+}
+
+static uint8_t bap_cp_notify(struct bt_conn *conn,
+			     struct bt_gatt_subscribe_params *params,
+			     const void *data, uint16_t length)
+{
+	struct bt_ascs_cp_rsp *rsp;
+	int i;
+
+	struct net_buf_simple buf;
+
+	BT_DBG("conn %p len %u", conn, length);
+
+	if (!data) {
+		BT_DBG("Unsubscribed");
+		params->value_handle = 0x0000;
+		return BT_GATT_ITER_STOP;
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)data, length);
+
+	if (buf.len < sizeof(*rsp)) {
+		BT_ERR("Control Point Notification too small");
+		return BT_GATT_ITER_STOP;
+	}
+
+	rsp = net_buf_simple_pull_mem(&buf, sizeof(*rsp));
+
+	for (i = 0; i < rsp->num_ase; i++) {
+		struct bt_ascs_cp_ase_rsp *ase_rsp;
+
+		if (buf.len < sizeof(*ase_rsp)) {
+			BT_ERR("Control Point Notification too small");
+			return BT_GATT_ITER_STOP;
+		}
+
+		ase_rsp = net_buf_simple_pull_mem(&buf, sizeof(*ase_rsp));
+
+		BT_DBG("op %s (0x%02x) id 0x%02x code %s (0x%02x) "
+		       "reason %s (0x%02x)", bt_ascs_op_str(rsp->op), rsp->op,
+		       ase_rsp->id, bt_ascs_rsp_str(ase_rsp->code),
+		       ase_rsp->code, bt_ascs_reason_str(ase_rsp->reason),
+		       ase_rsp->reason);
+	}
 
-int bap_config(struct bt_audio_stream *stream,
-	       struct bt_audio_capability *cap,
-	       struct bt_codec *codec)
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t bap_ep_notify(struct bt_conn *conn,
+			     struct bt_gatt_subscribe_params *params,
+			     const void *data, uint16_t length)
 {
-	struct bt_audio_ep *ep = stream->ep;
-	struct bt_ascs_config_op *op;
-	struct net_buf_simple *buf;
-	int err;
+	struct net_buf_simple buf;
+	struct bt_audio_ep *ep;
 
-	buf = bt_audio_ep_create_pdu(BT_ASCS_CONFIG_OP);
+	ep = CONTAINER_OF(params, struct bt_audio_ep, subscribe);
 
-	op = net_buf_simple_add(buf, sizeof(*op));
-	op->num_ases = 0x01;
+	BT_DBG("conn %p ep %p len %u", conn, ep, length);
 
-	err = bt_audio_ep_config(ep, buf, cap, codec);
-	if (err) {
-		return err;
+	if (!data) {
+		BT_DBG("Unsubscribed");
+		params->value_handle = 0x0000;
+		return BT_GATT_ITER_STOP;
 	}
 
-	err = bt_audio_ep_send(stream->conn, ep, buf);
-	if (err) {
-		return err;
+	net_buf_simple_init_with_data(&buf, (void *)data, length);
+
+	if (buf.len < sizeof(struct bt_ascs_ase_status)) {
+		BT_ERR("Notification too small");
+		return BT_GATT_ITER_STOP;
+	}
+
+	bap_ep_set_status(ep, &buf);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static int bap_ep_subscribe(struct bt_conn *conn, struct bt_audio_ep *ep)
+{
+	BT_DBG("ep %p handle 0x%02x", ep, ep->handle);
+
+	if (ep->subscribe.value_handle) {
+		return 0;
+	}
+
+	ep->subscribe.value_handle = ep->handle;
```

Comment:
Isn't that implied? Are there any other handles it could be?

Reply:
Connection handle, CIS handle some local generated EP-handle :-) No need to change, one will realize soon enough - had I read the code in the IDE rather than online, I would have found out in no time ;-)

Reply:
The function might get updated once we continue to update the code, so let's leave as is for now.